rfd/0080-yubikey-piv.md

```diff
@@ -0,0 +1,468 @@
+
+Would you like to store this key in PIN protected PIV metadata, so that
+tsh can find it automatically? This is not recommended if you plan to use
+the Management Key with other PIV applications, like `yubikey-agent`. (y/N):
+
+The Management key has been added to PIN protected metadata. You can 
+retrieve this key with `tsh piv configure --get-mgm-key` for use with
+other PIV clients, like `ykman` or `yubico-piv-tool`.
+
+> tsh login --piv
+Enter the PIV card's PIN:
+// Management key retrieved from metadata, so we don't prompt for it
+```
+
+This command will also be sharded into each of the three individual actions with `--set-pin`, `--set-puk`, and `--set-mgm-key`.
+
+If the user's PIN is blocked by PUK, then `tsh piv configure --set-pin` will notify the user and prompt them for PUK to reset their PIN instead.
+
+```
+> tsh piv configure --set-pin
+PIN is blocked, using PUK to reset PIN.
+Warning: three failures to enter the correct PUK will lock the PUK
+and you will be required to `tsh piv configure --reset` to reset your
+PIV card to defaults and erase all private key data.
+
+// Prompt user for current PUK and new PIN
+Enter the current PUK:
+Enter a new PIN:
+Repeat for confirmation:
+New PIN set.
+```
+
+We will also provide the user with the flags `--get-mgm-key` and `--reset`, which are noted in the command outputs above.
+
+```
+> tsh piv configure --get-mgm-key
+
+// Retrieve management key from metadata
+Management Key: <24 byte random code>
+WARNING: This is not recommended if you are using your PIV card with any other applications. Continue? (y/N):
+
+// Or if management key key not found in metadata...
+Management Key not found, try configuring it with `tsh piv configure --set-mgm-key`
+```
+
+```
+> tsh piv configure --reset
+// Reset PIV to default administrative keys and wipe private key and certificate data, equivalent to `ykman reset`
+```
+
+### UX
+
+#### new `tsh` flags
+
+| Name | Default Value(s) | Allowed Value(s) | Description |
+|-|-|-|-|
+| `--piv` | - | - | During login, use PIV to generate and store private key data for your login session. |
+| `--piv-slot` | 9a | 9a, 9c-9e, 82-95 | Configure which PIV slot is used is used for this login session. |
+| `--piv-yubikey-serial` | - | - | Specify which yubikey to use for PIV login by serial number. This is only needed when more than one smart card is connected. |
+| `--piv-pin-policy` | once | always once never | Can be provided with `tsh login -piv` to control how often you need to enter your PIV PIN to access the PIV slot's private key data. |
+| `--piv-touch-policy` | cached | always cached never | Can be provided with `tsh login -piv` to control how often you need to touch your PIV card to access the PIV slot's private key data. |
+| `--piv-overwrite` | - | - | When provided, `tsh login -piv` will overwrite a PIV slot's existing data if it exists. |
+
+### Additional considerations
+
+#### Kubernetes and MongoDB support
```

Comment:
@r0mant @jakule can you take a look at this section? Let me know if you have any workarounds in mind, and whether you need me to flush out a solution for the initial implementation.

Reply:
I think that we shouldn't worry about that at that point. Let's focus on SSH and when we decide to do the same for K8S and DBs we can write a new RFD.

Reply:
@Joerger @jakule `tsh db connect` probably won't work, not just for Mongo but for any database since the clients require key file, but would connecting through the local proxy work? `tsh proxy db --tunnel` - this is what Teleport Connect uses also under the hood.

I think we just need to make sure that `tsh proxy xxx` commands work with this, then any protocol will work through the tunnel.

Reply:
You're right, I just tested `tsh db connect redis` on https://github.com/gravitational/teleport/pull/15335, and it fails with `Could not negotiate a TLS connection: Invalid private key
Could not negotiate a TLS connection: Invalid private key`. But `tsh poxy db --tunnel` does work.